agent/xds/listeners.go

```diff
@@ -1068,14 +1056,48 @@ func (s *ResourceGenerator) makeTerminatingGatewayListener(
 		}
 	}
 
+	for _, svc := range cfgSnap.TerminatingGateway.ValidEndpoints() {
+		clusterName := connect.ServiceSNI(svc.Name, "", svc.NamespaceOrDefault(), svc.PartitionOrDefault(), cfgSnap.Datacenter, cfgSnap.Roots.TrustDomain)
+
+		intentions := cfgSnap.TerminatingGateway.Intentions[svc]
+		svcConfig := cfgSnap.TerminatingGateway.ServiceConfigs[svc]
+
+		cfg, err := ParseProxyConfig(svcConfig.ProxyConfig)
+		if err != nil {
+			// Don't hard fail on a config typo, just warn. The parse func returns
+			// default config if there is an error so it's safe to continue.
+			s.Logger.Warn(
+				"failed to parse Connect.Proxy.Config for linked service",
+				"service", svc.String(),
+				"error", err,
+			)
+		}
+
+		var endpoint *structs.EndpointConfig
+		if cfgSnap.TerminatingGateway.EndpointServices[svc].IsEndpoint {
+			endpoint = &svcConfig.Endpoint
+		}
+		clusterChain, err := s.makeFilterChainTerminatingGateway(cfgSnap, clusterName, svc, intentions, cfg.Protocol, endpoint)
+		if err != nil {
+			return nil, fmt.Errorf("failed to make filter chain for cluster %q: %v", clusterName, err)
+		}
+		l.FilterChains = append(l.FilterChains, clusterChain)
+	}
+
 	// Before we add the fallback, sort these chains by the matched name. All
 	// of these filter chains are independent, but envoy requires them to be in
 	// some order. If we put them in a random order then every xDS iteration
 	// envoy will force the listener to be replaced. Sorting these has no
 	// effect on how they operate, but it does mean that we won't churn
 	// listeners at idle.
 	sort.Slice(l.FilterChains, func(i, j int) bool {
-		return l.FilterChains[i].FilterChainMatch.ServerNames[0] < l.FilterChains[j].FilterChainMatch.ServerNames[0]
+		if len(l.FilterChains[i].FilterChainMatch.PrefixRanges) > 0 && len(l.FilterChains[j].FilterChainMatch.PrefixRanges) > 0 {
+			return l.FilterChains[i].FilterChainMatch.PrefixRanges[0].AddressPrefix < l.FilterChains[j].FilterChainMatch.PrefixRanges[0].AddressPrefix
+		}
+		if len(l.FilterChains[i].FilterChainMatch.ServerNames) > 0 && len(l.FilterChains[j].FilterChainMatch.ServerNames) > 0 {
+			return l.FilterChains[i].FilterChainMatch.ServerNames[0] < l.FilterChains[j].FilterChainMatch.ServerNames[0]
+		}
+		return false
```

Comment:
I changed this to use string compare. It's less generic but preserve the order. Also, I think we had a bug in here. If serverNames were empty it would panic.